Use new classified listing categories

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Generalization

## Description

This is the second PR in the process of generalizing classified listings. PR #7172 introduced a model, this PR starts the cleanup process and switches our code to use it. Since this is live already, I'm trying to do this carefully, i.e. most methods in `ClassifiedListing` that wrapped the constant are still there and access the DB instead. We can refactor this in a follow-up PR.

## Related Tickets & Documents

#6740 

## Added tests?

- [X] updated existing one

## Added to documentation?

- [X] no documentation needed

## What gif best describes this PR or how it makes you feel?

![image](https://user-images.githubusercontent.com/47985/79534258-46db7200-80a4-11ea-8852-6d7f33335c43.png)

citizen428 commented on app/javascript/listings/elements/categories.jsx:
```diff
@@ -19,7 +19,7 @@ class Categories extends Component {
   details = () => {
     const { categoriesForDetails } = this.props;
-    const rules = categoriesForDetails.map(category => {
+    const rules = categoriesForDetails.map((category) => {
```
Prettier strikes again.

citizen428 commented on app/javascript/packs/listingForm.jsx:
```diff
@@ -9,7 +9,7 @@ function loadElement() {
       categoriesForSelect,
       categoriesForDetails,
-    } = root.dataset;  
+    } = root.dataset;
```
All changes in this file courtesy your friendly neighborhood Prettier.

citizen428 commented on app/models/classified_listing.rb:
```diff
@@ -4,22 +4,9 @@ class ClassifiedListing < ApplicationRecord
-
   attr_accessor :action
 
+  belongs_to :classified_listing_category
```
This is no longer optional (see #7172)

citizen428 commented on app/models/classified_listing.rb:
```diff
@@ -40,34 +27,38 @@ class ClassifiedListing < ApplicationRecord
-
-  def author
-    organization || user
+  # TODO: refactor this class method block
```
In order to keep the number of moving parts in each PR relatively low I decided to keep the old interface and just swap out the implementation.

citizen428 commented on app/models/classified_listing.rb:
```diff
@@ -40,34 +27,38 @@ class ClassifiedListing < ApplicationRecord
+    end.deep_symbolize_keys
+  end
+
+  def category
```
This new method is here for compatibility reasons with the old implementation.

citizen428 commented on app/models/classified_listing.rb:
```diff
@@ -102,11 +93,7 @@ def validate_tags
-
   def create_slug
-    self.slug = title.to_s.downcase.parameterize.tr("_", "") + "-" + rand(100_000).to_s(26)
+    self.slug = "#{title.downcase.parameterize.delete('_')}-#{rand(100_000).to_s(26)}"
```
Simplified slug generation code to use interpolation instead of `String#+`.

citizen428 commented on app/models/classified_listing_category.rb:
```diff
@@ -0,0 +1,6 @@
+class ClassifiedListingCategory < ApplicationRecord
```
This won't be in the diff once #7172 was merged.

citizen428 commented on config/routes.rb:
```diff
@@ -47,11 +47,10 @@
     resources :broadcasts, only: %i[index new create edit update]
     resources :buffer_updates, only: %i[create update]
-    resources :classified_listings, only: %i[index edit update destroy]
+    resources :classified_listings, only: %i[index edit update destroy], alias: :listings
```
This seems much easier than we had before and should achieve the same.

citizen428 commented on db/migrate/20200409043946_create_classified_listing_categories.rb:
```diff
@@ -0,0 +1,14 @@
+class CreateClassifiedListingCategories < ActiveRecord::Migration[5.2]
```
This won't be in the diff once #7172 was merged.

citizen428 commented on db/migrate/20200409050122_add_listing_category_to_classified_listing.rb:
```diff
@@ -0,0 +1,10 @@
+class AddListingCategoryToClassifiedListing < ActiveRecord::Migration[5.2]
```
This won't be in the diff once #7172 was merged.

citizen428 commented on db/schema.rb:
```diff
@@ -10,8 +10,7 @@
 
-ActiveRecord::Schema.define(version: 2020_04_07_091449) do
-
+ActiveRecord::Schema.define(version: 2020_04_09_050122) do
```
This won't be in the diff once #7172 was merged.

citizen428 commented on lib/tasks/temporary/classified_listings.rake:
```diff
@@ -0,0 +1,13 @@
+namespace :temporary do
```
This won't be in the diff once #7172 was merged.

citizen428 commented on spec/factories/classified_listing_categories.rb:
```diff
@@ -0,0 +1,12 @@
+FactoryBot.define do
```
This won't be in the diff once #7172 was merged.

citizen428 commented on spec/models/classified_listing_category_spec.rb:
```diff
@@ -0,0 +1,16 @@
+require "rails_helper"
```
This won't be in the diff once #7172 was merged.

citizen428 commented:
OMG, this is ready for review. I'd appreciate if people could download the branch and try things out (I did as both user and admin), as there are quite a few changes. Though as mentioned in the description, I tried to keep most of the API intact for now, at the cost of potentially complicating some code for a short time until the 3rd PR lands.

rhymes commented on app/controllers/api/v0/classified_listings_controller.rb:
```diff
@@ -16,10 +16,13 @@ class ClassifiedListingsController < ApiController
+          includes(:user, :organization, :taggings, :classified_listing_category)
 
+        if params[:category].present?
+          category = ClassifiedListingCategory.find_by(slug: params[:category])
```
if the category is not found, it will return all listings, is that intended?

citizen428 replied:
That's the same behavior as before:

```ruby
@classified_listings = @classified_listings.where(category: params[:category]) if params[:category].present?
```

 `params[:category]` will only be present of we filter.

rhymes commented on app/controllers/concerns/classified_listings_toolkit.rb:
```diff
@@ -13,11 +13,14 @@ def publish_listing
     @classified_listing.save
   end
 
+  # TODO: why not just @classified_listing.update(listing_params)?
```
no, but you could use `assign_attributes` not to trigger an update at this stage I guess

rhymes replied:
ps, don't forget to add the name of the team, it's clearer for onlookers :)

something like `# TODO: [thepracticaldev/oss]`

citizen428 replied:
I had `assign_attributes` locally for a while, it seemed to work fine. I reverted it before pushing though, because this is idiosyncratic enought that I figured there must be context. Anyway, I'll jut change it back.

Re the TODO: I can change it, but given that I wanted to start on the next PR within a few days, the comments should be around too long anyway.

mstruve replied:
My guess would be that some of these params might come through as nil or blank when not changed and this prevents them from being incorrectly set. That is totally just a guess. You could still refactor though with something like
```ruby
listing_params.keep_if{|k, v| v.present?}
@classified_listing.assign_attributes(listing_params)
```

citizen428 replied:
@mstruve Was thinking the same, but the way it's currently implemented only protects against `nil` and not blank, so I was a bit reluctant to change it before at least double-checking if someone knows more. 

Anyway, I've decided to leave this method as-is for now and keep the changes for the 3rd PR.

rhymes commented on app/controllers/social_previews_controller.rb:
```diff
@@ -51,6 +51,7 @@ def comment
 
   private
 
+  # TODO: don't hardcode this
```
same note about `# TODO: [thepracticaldev/oss]`

rhymes commented on app/models/classified_listing.rb:
```diff
@@ -39,41 +24,45 @@ class ClassifiedListing < ApplicationRecord
+    classified_listing_category&.slug
+  end
+
+  def cost
```
🔥

mstruve replied:
Should we cache this to avoid the lookup multiple times?
```ruby
@cost ||= classified_listing_category&.cost || ClassifiedListingCategory.select(:cost).find_by(slug: category)&.cost
```

citizen428 replied:
Where we use it (mostly in `ClassifiedListingsToolkit`) it's generally assigned to a variable that will be checked. I can add it though. The only potential problem is that if we change the price anyone who started editing a listing just before will spend the old amount, but that's unlikely I guess.

rhymes commented on db/schema.rb:
```diff
@@ -322,6 +322,17 @@
     t.index ["slug"], name: "index_classified_listing_categories_on_slug", unique: true
   end
 
+  create_table "classified_listing_categories", force: :cascade do |t|
```
isn't this already present on master?

citizen428 replied:
Yeah, my local branch must have gotten messed up during a rebase or something. Thanks!

rhymes commented on spec/requests/listings_spec.rb:
```diff
@@ -3,8 +3,20 @@
+ end
 
-  describe "GETS /listings" do
+    describe "GETS /listings" do
```
too much space will kill you :D

citizen428 commented:
Thanks for the feedback @rhymes and @mstruve :smiley:

1. Categories added to seeds.
2. Added the `includes` for the eager loading, but then Bullet started failing with unused eager-loading for the case where we redirect to the internal view for moderators.  I think I fixed it, but if I didn't it will probably have to wait for the next PR, because this one has been dragging on too much already.
3. `cost` is now cached in  an instance variable.
4. Will you create an issue for the last thing you mentioned @rhymes?

rhymes commented:
> Will you create an issue for the last thing you mentioned @rhymes?

ok!

citizen428 commented:
Hm, I created a subset. Why would you like all of them locally? In fact given the current generalization efforts I was tempted to add three completely random ones at first.

citizen428 commented:
@mstruve A classic case of "it works better for others if you push it to remote" :wink: I'll leave it for the follow-up PR.
